### PR TITLE
Initial implementation of async storage support for DevWorkspaces

### PR DESCRIPTION
**Pull request by amisevsk**
### What does this PR do?
This PR adds support for async storage, defined by setting the label `controller.devfile.io/storage-type: async` on a DevWorkspace. When async storage is used, the same mechanism as in Eclipse Che is deployed:
1. A configmap (for the storage server) and secret (for the storage sidecar) are created. These contain an ssh keypair
2. A deployment for the async storage server is created, along with a service
3. A sidecar is added to the workspace pod that syncs data to the async storage server.

This PR also refactors how storage is provisioned/cleaned up in general, collecting all functionality into a new interface. The first two commits are mostly refactors of internal functionality, to move all PVC-related steps into this interface for the common PVC strategy. This interface will also make it fairly easy to add support for `ephemeral` as a storage type.

Async storage currently comes with some caveats:
1. We only allow one DevWorkspace to run using async storage at a time (per-namespace), additional ones fail startup.
2. We can't delete workspaces while there's another async workspace running (as the common PVC is mounted and in-use)
3. No tests are implemented yet :upside_down_face: 

There are a lot of gaps and small edge cases that aren't handled elegantly (and probably even more that I'm not aware of yet), but as an initial implementation it *mostly* works. 

One important bug that requires additional work is that configmaps mounted via `subPath` do not propagate changes, after the first workspace is created, addtional workspaces cannot sync data. Fixing this would likely require changing the async server image.

### What issues does this PR fix or reference?
Closes https://github.com/devfile/devworkspace-operator/issues/267

### Is it tested? How?
This PR adds a yaml file: `samples/with-k8s-ref/async-theia-next.yaml`, which is the usual theia-next sample with async storage enabled. This can be used to test basic use cases (create workspace, change a file, stop workspace, restart workspace -- note it takes a second for the files to repopulate after startup). 

To test cleanup, it's necessary to ssh into either a dedicated pod or the async storage server and check the contents of `/async-storage`. A flow that does this is
```bash
kubectl apply -f samples/with-k8s-ref/async-theia-next.yaml
# wait started; change file; stop workspace; wait till workspace pods terminated
kubectl exec -it async-storage-xxxxx -- /bin/sh
# ls -al /async-storage, look at files
yq -y '.metadata.name = "alt-theia-next"' samples/with-k8s-ref/async-theia-next.yaml | kubectl apply -f -
# note: this image won't be able to sync due to issue described above
kubectl patch dw alt-theia-next --type merge -p '{"spec": {"started": false}}'
kubectl delete dw theia # Delete initial workspace
kubectl exec -it async-storage-xxxxx -- ls -al /async-storage # should be empty
kubectl delete dw alt-theia-next
# should be no async-storage deployment left on cluster after all workspaces deleted.
```

### CODE REVIEW

File: controllers/workspace/devworkspace_controller.go [hunk 1]
```diff
@@ -190,25 +191,35 @@ func (r *DevWorkspaceReconciler) Reconcile(req ctrl.Request) (reconcileResult ct
...
 		reqLogger.Info("DevWorkspace start failed")
 		reconcileStatus.Phase = devworkspace.WorkspaceStatusFailed
-		reconcileStatus.Conditions[devworkspace.WorkspaceFailedStart] = fmt.Sprintf("Error processing devfile volumes: %s", err)
+		reconcileStatus.Conditions[devworkspace.WorkspaceFailedStart] = fmt.Sprintf("Error provisioning storage: %s", err)
 		return reconcile.Result{}, nil
 	}
+	err = storageProvisioner.ProvisionStorage(devfilePodAdditions, workspace, clusterAPI)
+	if err != nil {
+		switch storageErr := err.(type) {
+		case *storagelib.NotReadyError:
```
[the remark this answers is not in the document]
Comment (amisevsk) [reply controllers/workspace/devworkspace_controller.go:205]: yes :D 

I know I did this in _one_ of my PRs, but maybe the change was lost in a rebase.

File: pkg/provision/storage/asyncstorage/configuration.go [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+//
+// If the k8s objects do not exist, an SSH keypair is generated and a secret and configmap are created on the cluster.
+// This function works on two streams:
+// 1. If the async storage SSH secret for the given workspace does not exist on the cluster, an SSH keypair are generated, a
+//    Secret is synced to the cluster and the corresponding authorized key is added to the ConfigMap
+// 2. If the async storage SSH secret exists, its content is read, and the ConfigMap is verified to contain the corresponding public
+//    key in authorized_keys.
+// In both cases, if the ConfigMap does not exist, it is created.
+//
+// Returns SSHConfigNotReadyError if changes were made to the cluster.
```
[the remark this answers is not in the document]
Comment (amisevsk) [reply pkg/provision/storage/asyncstorage/configuration.go]: Good catch, fixed.

File: controllers/workspace/finalize.go [hunk 3]
```diff
@@ -79,163 +55,59 @@ func (r *DevWorkspaceReconciler) finalize(ctx context.Context, log logr.Logger,
...
-		},
-	}
-
-	err := controllerutil.SetControllerReference(workspace, job, r.Scheme)
-	if err != nil {
-		return nil, err
-	}
-	return job, nil
+	log.Info("PVC clean up successful; clearing finalizer")
+	coputil.RemoveFinalizer(workspace, pvcCleanupFinalizer)
```
[the remark this answers is not in the document]
Comment (amisevsk) [reply controllers/workspace/finalize.go]: Done.

File: controllers/workspace/devworkspace_controller.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (amisevsk) [reply controllers/workspace/devworkspace_controller.go]: Re-done :)

File: pkg/provision/storage/asyncStorage.go [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+	// Note we need to check phase so as to not accidentally fail an already-running workspace when a second one
+	// is created.
+	if numWorkspaces > 1 && workspace.Status.Phase != dw.WorkspaceStatusRunning {
+		return &ProvisioningError{
+			Message: fmt.Sprintf("Cannot provision storage for workspace %s", workspace.Name),
+			Err:     fmt.Errorf("at most one workspace using async storage can be running in a namespace"),
+		}
+	}
+
+	// Add ephemeral volumes
```
[the remark this answers is not in the document]
Comment (amisevsk) [reply pkg/provision/storage/asyncStorage.go:58]: Good point -- makes sense. Thanks!

Applied the changes. Now shared.go contains functions
```go 
func addEphemeralVolumesFromWorkspace(workspace *dw.DevWorkspace, podAdditions *v1alpha1.PodAdditions) error

func addEphemeralVolumesToPodAdditions(podAdditions *v1alpha1.PodAdditions, workspaceVolumes []dw.Component) (addedVolumes []corev1.Volume, err error)
```
which may be a little confusing. 

I wonder if there's a way to clean this up, given that we have to handle the projects volume semi-delicately.

File: pkg/provision/storage/asyncstorage/sidecar.go [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+		Ports: []corev1.ContainerPort{
+			{
+				ContainerPort: 4445,
+				Protocol:      corev1.ProtocolTCP,
+			},
+		},
+		Env: []corev1.EnvVar{
+			{
+				Name:  "RSYNC_PORT",
+				Value: "2222",
```
[the remark this answers is not in the document]
Comment (amisevsk) [reply pkg/provision/storage/asyncstorage/sidecar.go]: Good catch :+1:

File: controllers/workspace/finalize.go [hunk 6]
```diff
@@ -14,50 +14,26 @@ package controllers
...
-	pvcCleanupPodMemoryRequest = resource.MustParse(constants.PVCCleanupPodMemoryRequest)
-	pvcCleanupPodCPULimit      = resource.MustParse(constants.PVCCleanupPodCPULimit)
-	pvcCleanupPodCPURequest    = resource.MustParse(constants.PVCCleanupPodCPURequest)
+	storageCleanupFinalizer = "storage.controller.devfile.io"
 )
 
 func (r *DevWorkspaceReconciler) finalize(ctx context.Context, log logr.Logger, workspace *devworkspace.DevWorkspace) (reconcile.Result, error) {
-	if !coputil.HasFinalizer(workspace, pvcCleanupFinalizer) {
+	if !coputil.HasFinalizer(workspace, storageCleanupFinalizer) {
 		return reconcile.Result{}, nil
```
[the remark this answers is not in the document]
Comment (sleshchenko) [reply controllers/workspace/finalize.go:37]: I debugged the issue sometimes, 

`r.Client.Status().Update(ctx, workspace)` or `r.Client.Status().Update(ctx, workspace)` has not affect... And then I realized... It's the URL in the `kubectl get dw`... It's old CRD... OH MY GOD!!!

It should not happen after my fixed, I just yesterday played with old CRDs and forget to update_crd after it. We definitely need to indicate some info in the git repo, which DevWorkspace CRDs are initialized into https://github.com/devfile/devworkspace-operator/tree/main/deploy/templates/crd/bases. But it's another story.

**Review by sleshchenko — COMMENTED** [review]
I haven't finished testing yet.

**Review by sleshchenko — COMMENTED** [review]
After I commented everything works seems to work.

Does PR has actual instructions to test?
```
kubectl delete dw theia # Delete initial workspace
kubectl exec -it async-storage-xxxxx -- ls -al /async-storage # should be empty
```
^ isn't async-storage supposed to be scaled to 0 after DW is removed?

File: controllers/workspace/finalize.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (sleshchenko) [reply controllers/workspace/finalize.go:37]: A dedicated issue is created to expose terminating state:

DevWorkspace should have a dedicated phase for terminating https://github.com/devfile/devworkspace-operator/issues/333

[the remark this answers is not in the document]
Comment (sleshchenko) [reply controllers/workspace/finalize.go:37]: Solving since it's just a noise in the current PR scope, but still may be interesting to read.

**Review by sleshchenko — APPROVED** [review]
Such a big piece of work! It's a great initial implementation :muscle: 

Nothing is blocking apart from introducing WAIT_BOUND PVC, which does not work by default on RHPDS cluster.

Approving since I think everybody will win from merging this PR (after commenting WAIT_BOUND) and fixing left comments in a separate PR.
But it's up to you.

**Review by JPinkney — APPROVED** [review]
Approving as is as well, since this PR is so complicated (and to avoid another rebase) we should probably just merge and fix comments later

File: pkg/provision/storage/asyncstorage/sidecar.go [hunk 7]
```diff
@@ -51,7 +51,7 @@ func GetAsyncSidecar(sshVolumeName string, volumes []corev1.Volume) *corev1.Cont
 		Env: []corev1.EnvVar{
 			{
 				Name:  "RSYNC_PORT",
-				Value: "2222",
+				Value: rsyncPortStr,
```
[the remark this answers is not in the document]
Comment (amisevsk) [reply pkg/provision/storage/asyncstorage/sidecar.go:54]: :+1:

File: controllers/workspace/devworkspace_controller.go [hunk 8]
```diff
@@ -192,25 +193,35 @@ func (r *DevWorkspaceReconciler) Reconcile(req ctrl.Request) (reconcileResult ct
 		reconcileStatus.Conditions[devworkspace.WorkspaceFailedStart] = fmt.Sprintf("Error processing devfile: %s", err)
 		return reconcile.Result{}, nil
 	}
-	err = storagelib.RewriteContainerVolumeMounts(workspace.Status.WorkspaceId, devfilePodAdditions, workspace.Spec.Template)
+
+	storageProvisioner, err := storage.GetProvisioner(workspace)
```
[the remark this answers is not in the document]
Comment (amisevsk) [reply controllers/workspace/devworkspace_controller.go:197]: Pushing this to a follow-up PR for easier review.

File: controllers/workspace/finalize.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (amisevsk) [reply controllers/workspace/finalize.go:37]: :+1: thanks, good catch

File: pkg/provision/storage/asyncStorage.go [hunk 9]
```diff
@@ -0,0 +1,251 @@
...
+	case 0:
+		// no problem
+	case 1:
+		if workspace.Spec.Started {
+			// This is the only workspace using the async server, we can safely stop it
+			break
+		}
+		// Another async workspace is currently running; we can't safely clean up
+		return &ProvisioningError{
+			Message: "Cannot cleanup storage",
```
[the remark this answers is not in the document]
Comment (amisevsk) [reply pkg/provision/storage/asyncStorage.go:162]: :+1:

File: controllers/workspace/provision/object.go [hunk 10]
```diff
@@ -91,9 +91,9 @@ func SyncObject(object runtime.Object, client client.Client, reqLogger logr.Logg
 		updateErr := client.Update(context.TODO(), object)
 		if errors.IsConflict(updateErr) {
 			fmt.Println("Suppressing conflict in ", objType.String())
-			return true, nil
+			return found, true, nil
 		}
-		return true, updateErr
+		return object, true, updateErr
```
[the remark this answers is not in the document]
Comment (amisevsk) [reply controllers/workspace/provision/object.go:96]: I agree, this could use some cleaning up. In general, we suppress conflict errors because they're semi-common and have no effect on functionality (the resolution is to just try again, since we have an out-of-date resource we're working on).

I'll push this fix to the follow-up PR for easier review.

File: pkg/provision/storage/shared.go [hunk 11]
```diff
@@ -0,0 +1,192 @@
...
+	if didChange {
+		return nil, &NotReadyError{
+			Message: "Updated common PVC on cluster",
+		}
+	}
+	currPVC, ok := currObject.(*corev1.PersistentVolumeClaim)
+	if !ok {
+		return nil, errors.New("tried to sync PVC to cluster but did not get a PVC back")
+	}
+	if currPVC.Status.Phase != corev1.ClaimBound {
```
[the remark this answers is not in the document]
Comment (amisevsk) [reply pkg/provision/storage/shared.go:99]: This is what I get for not testing on a real cluster :). Commented out for now.

File: controllers/workspace/devworkspace_controller.go [hunk 12]
```diff
@@ -192,25 +193,35 @@ func (r *DevWorkspaceReconciler) Reconcile(req ctrl.Request) (reconcileResult ct
...
 		reconcileStatus.Phase = devworkspace.WorkspaceStatusFailed
-		reconcileStatus.Conditions[devworkspace.WorkspaceFailedStart] = fmt.Sprintf("Error processing devfile volumes: %s", err)
+		reconcileStatus.Conditions[devworkspace.WorkspaceFailedStart] = fmt.Sprintf("Error provisioning storage: %s", err)
 		return reconcile.Result{}, nil
 	}
+	err = storageProvisioner.ProvisionStorage(devfilePodAdditions, workspace, clusterAPI)
+	if err != nil {
+		switch storageErr := err.(type) {
+		case *storage.NotReadyError:
+			reqLogger.Info(storageErr.Message)
```
[the remark this answers is not in the document]
Comment (amisevsk) [reply controllers/workspace/devworkspace_controller.go:208]: :+1:

This is tricky though -- our conditions are generally e.g. `componentsReady`, etc. so propagating to a `storageReady` condition doesn't quite make sense (the condition being `true` means that storage is ready). The correct solution would likely be adding support for setting conditions to `false` (as opposed to unknown), but that will have to be a separate PR.

I've created issue https://github.com/devfile/devworkspace-operator/issues/337 to track this.

File: pkg/provision/storage/asyncStorage.go [hunk 13]
```diff
@@ -0,0 +1,251 @@
...
+
+func (p *AsyncStorageProvisioner) ProvisionStorage(podAdditions *v1alpha1.PodAdditions, workspace *dw.DevWorkspace, clusterAPI provision.ClusterAPI) error {
+	numWorkspaces, _, err := p.getAsyncWorkspaceCount(clusterAPI)
+	if err != nil {
+		return err
+	}
+	// If there is more than one started workspace using async storage, then we fail starting additional ones
+	// Note we need to check phase so as to not accidentally fail an already-running workspace when a second one
+	// is created.
+	if numWorkspaces > 1 && workspace.Status.Phase != dw.WorkspaceStatusRunning {
```
[the remark this answers is not in the document]
Comment (amisevsk) [reply pkg/provision/storage/asyncStorage.go:51]: Part of the issue here is that I don't want to _actually_ restrict to 1 running workspace at a time -- the only reason we can't run multiple workspaces simultaneously is that cleanup becomes hard (we've no way to clean up the common PVC while it's mounted to the async server). 

The idea behind this snippet is to fail workspaces whenever we see multiple async workspaces in a namespace. To avoid failing an already-started workspace when a second one is created, we skip workspaces that are already Running, which _should_ (my logic might be wrong here) lead to the cases:
1. Start an async workspace -> starts fine
2. Start two+ async workspaces at same time -> both fail
3. Start a second async workspace when one is already running -> new workspace fails but running workspace is unaffected.

The second case is tricky, since it becomes a race to see which one starts first, and then we have to fail the second. It's safer to just fail hard as soon as possible. Ideally whoever is creating devworkspaces limits it to one per namespace for now (e.g. che dashboard). Since this check happens on reconciles, the only way an async workspace enters the `running` state (which requires a reconcile) is if it is the only async workspace on the cluster at the time that reconcile is triggered. 

The main gap I can see here is that I don't think we properly clean resources when a workspace fails. If a workspace fails after the deployment has been created, that deployment will stick around for now. 

Created issue https://github.com/devfile/devworkspace-operator/issues/338 to partially look at this.

File: pkg/provision/storage/asyncstorage/deployment.go [hunk 14]
```diff
@@ -0,0 +1,157 @@
...
+	replicas := int32(1)
+	terminationGracePeriod := int64(1)
+	modeReadOnly := int32(416) // 0640 (octal) in base-10
+
+	deployment := &appsv1.Deployment{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "async-storage", // TODO
+			Namespace: namespace,
+			Labels: map[string]string{
+				"app.kubernetes.io/name":    "async-storage", // TODO
```
[the remark this answers is not in the document]
Comment (amisevsk) [reply pkg/provision/storage/asyncstorage/deployment.go:62]: :D Thought I had caught them all but I guess not. Will fix.

File: pkg/provision/storage/asyncstorage/deployment.go [hunk 15]
```diff
@@ -0,0 +1,157 @@
...
+			},
+			Strategy: appsv1.DeploymentStrategy{
+				Type: appsv1.RecreateDeploymentStrategyType,
+			},
+			Template: corev1.PodTemplateSpec{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "async-storage-server",
+					Namespace: namespace,
+					Labels: map[string]string{
+						"controller.devfile.io/component": "async-storage",
```
[the remark this answers is not in the document]
Comment (amisevsk) [reply pkg/provision/storage/asyncstorage/deployment.go:81]: :+1:

File: pkg/provision/storage/asyncstorage/deployment.go [hunk 16]
```diff
@@ -0,0 +1,157 @@
...
+					},
+				},
+				Spec: corev1.PodSpec{
+					Containers: []corev1.Container{
+						{
+							Name:  "async-storage-server",
+							Image: "quay.io/eclipse/che-workspace-data-sync-storage:0.0.1",
```
[the remark this answers is not in the document]
Comment (amisevsk) [reply pkg/provision/storage/asyncstorage/deployment.go:88]: Extracted these into `RELATED_IMAGE_*` env vars like we do for PVC cleanup.

File: pkg/provision/storage/doc.go [hunk 17]
```diff
@@ -0,0 +1,23 @@
...
+//
+// Contributors:
+//   Red Hat, Inc. - initial API and implementation
+//
+
+// Package storage contains library functions for provisioning volumes and volumeMounts in containers according to the
+// volume components in a devfile. These functions also handle mounting project sources to containers that require it.
+//
+// TODO:
+// - Add functionality for generating PVCs with the appropriate size based on size requests in the devfile
```
[the remark this answers is not in the document]
Comment (amisevsk) [reply pkg/provision/storage/doc.go:17]: No, but it's worth noting that we can't handle the size field in a Volume :)

**Comment by amisevsk** [comment]
Pushed fixes for relatively minor issues (error messages, constants, images used); I'll merge the PR in its current state and start prepping a PR to follow up on bigger comments.

File: pkg/provision/storage/asyncstorage/errors.go [hunk 18]
```diff
@@ -0,0 +1,17 @@
...
+//
+// Contributors:
+//   Red Hat, Inc. - initial API and implementation
+//
+
+package asyncstorage
+
+import "errors"
+
+var NotReadyError = errors.New("async storage component is not ready")
```
[the remark this answers is not in the document]
Comment (amisevsk) [reply pkg/provision/storage/asyncstorage/errors.go:17]: The error value here is just to distinguish between an unexpected error and a routine error (i.e. where we just need to requeue/wait). This is different from the `NotReadyError` in the storage package, which has additional context (like message and requeue duration). The storage package `NotReadyError` is part of the interface as it's read by the caller to determine information about how to requeue and set messages. 

The way to remove the three checks would be break out storage errors into a separate package, which is IMO messier than just doing standard golang copy-paste error handling. We could set the message and requeue delay in the asyncstorage package, but it would mean a circular import by default.

[the remark this answers is not in the document]
Comment (sleshchenko) [reply pkg/provision/storage/asyncstorage/errors.go:17]: I'm OK with keeping as is for time being.
